File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No80RemoveDuplicatesFromSortedArrayII.java
package com.hazza.LeetCodeByJava.SegmentTwo;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-09-02
 * Time: 11:08 AM
 */
public class No80RemoveDuplicatesFromSortedArrayII {
    // solution 1
    public int removeDuplicates(int[] nums) {
        int count = 1, duplicates = 1, len = nums.length;
        for (int i = 1; i < len; i++) {
            if (nums[i] == nums[i - 1]) {
                ++duplicates;
                if (duplicates <= 2) nums[count++] = nums[i];
            }
            else {
                nums[count++] = nums[i];
                duplicates = 1;
            }
        }

        return count;
    }

    // solution 2
    public int removeDuplicates2(int[] nums) {
        int count = 0;
        for (int n: nums)
            if (count < 2 || n > nums[count - 2])
                nums[count++] = n;

        return count;
    }
}

File: src/main/python/leetcode_by_python/SegmentTwo/No80_Remove_Duplicates_from_Sorted_Array_II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-07-15
# https://leetcode.com/problems/remove-duplicates-from-sorted-array-ii/
from typing import List


class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        cnt = 0

        for n in nums:
            if cnt < 2 or n > nums[cnt - 2]:
                nums[cnt] = n
                cnt += 1

        return cnt


if __name__ == '__main__':
    solution = Solution()

    assert solution.removeDuplicates([1, 1, 1, 2, 2, 3]) == 5
    assert solution.removeDuplicates([0, 0, 1, 1, 1, 1, 2, 3, 3]) == 7
